File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/python/algorithms/searching/red_black_tree.py
RED = "RED"
BLACK = "BLACK"


class Node:
    def __init__(self, value: int, color: str = RED) -> None:
        self.value = value
        self.color = color
        self.left: Node | None = None
        self.right: Node | None = None
        self.parent: Node | None = None


class RedBlackTree:
    def __init__(self) -> None:
        self.nil = Node(0, BLACK)
        self.root: Node = self.nil

    def insert(self, value: int) -> None:
        node = Node(value)
        node.left = self.nil
        node.right = self.nil

        parent = self.nil
        current = self.root
        while current != self.nil:
            parent = current
            current = current.left if node.value < current.value else current.right

        node.parent = parent
        if parent == self.nil:
            self.root = node
        elif node.value < parent.value:
            parent.left = node
        else:
            parent.right = node

        node.color = RED
        self._fix_insert(node)

    def contains(self, value: int) -> bool:
        current = self.root
        while current != self.nil:
            if value == current.value:
                return True
            current = current.left if value < current.value else current.right
        return False

    def inorder(self) -> list[int]:
        result: list[int] = []
        self._inorder(self.root, result)
        return result

    def _fix_insert(self, node: Node) -> None:
        while node.parent is not None and node.parent.color == RED:
            grandparent = node.parent.parent
            if grandparent is None:
                break
            if node.parent == grandparent.left:
                uncle = grandparent.right
                if uncle is not None and uncle.color == RED:
                    node.parent.color = BLACK
                    uncle.color = BLACK
                    grandparent.color = RED
                    node = grandparent
                else:
                    if node == node.parent.right:
                        node = node.parent
                        self._left_rotate(node)
                    node.parent.color = BLACK
                    grandparent.color = RED
                    self._right_rotate(grandparent)
            else:
                uncle = grandparent.left
                if uncle is not None and uncle.color == RED:
                    node.parent.color = BLACK
                    uncle.color = BLACK
                    grandparent.color = RED
                    node = grandparent
                else:
                    if node == node.parent.left:
                        node = node.parent
                        self._right_rotate(node)
                    node.parent.color = BLACK
                    grandparent.color = RED
                    self._left_rotate(grandparent)
        self.root.color = BLACK

    def _left_rotate(self, node: Node) -> None:
        right = node.right
        if right is None:
            return
        node.right = right.left
        if right.left != self.nil:
            right.left.parent = node
        right.parent = node.parent
        if node.parent == self.nil:
            self.root = right
        elif node == node.parent.left:
            node.parent.left = right
        else:
            node.parent.right = right
        right.left = node
        node.parent = right

    def _right_rotate(self, node: Node) -> None:
        left = node.left
        if left is None:
            return
        node.left = left.right
        if left.right != self.nil:
            left.right.parent = node
        left.parent = node.parent
        if node.parent == self.nil:
            self.root = left
        elif node == node.parent.right:
            node.parent.right = left
        else:
            node.parent.left = left
        left.right = node
        node.parent = left

    def _inorder(self, node: Node, result: list[int]) -> None:
        if node == self.nil:
            return
        self._inorder(node.left, result)
        result.append(node.value)
        self._inorder(node.right, result)


if __name__ == "__main__":
    sample = [42, 7, 19, 3, 99, 1]
    target = 19
    tree = RedBlackTree()
    for value in sample:
        tree.insert(value)

    print("Red-Black Tree")
    print(f"input  : {sample}")
    print(f"inorder: {tree.inorder()}")
    print(f"target : {target}")
    print(f"found  : {tree.contains(target)}")

File: 29Items-Projects/7Algorithms/1-Sorting-and-Searching/src/java/com/sortingsearching/searching/RedBlackTree.java
package com.sortingsearching.searching;

import java.util.ArrayList;
import java.util.List;

public final class RedBlackTree {
    private static final boolean RED = true;
    private static final boolean BLACK = false;

    private final Node nil = new Node(0, BLACK);
    private Node root = nil;

    public void insert(int value) {
        Node node = new Node(value, RED);
        node.left = nil;
        node.right = nil;

        Node parent = nil;
        Node current = root;
        while (current != nil) {
            parent = current;
            current = node.value < current.value ? current.left : current.right;
        }

        node.parent = parent;
        if (parent == nil) {
            root = node;
        } else if (node.value < parent.value) {
            parent.left = node;
        } else {
            parent.right = node;
        }

        fixInsert(node);
    }

    public boolean contains(int value) {
        Node current = root;
        while (current != nil) {
            if (value == current.value) {
                return true;
            }
            current = value < current.value ? current.left : current.right;
        }
        return false;
    }

    public List<Integer> inorder() {
        List<Integer> result = new ArrayList<>();
        inorder(root, result);
        return result;
    }

    private void fixInsert(Node node) {
        while (node.parent != null && node.parent.color == RED) {
            Node grandparent = node.parent.parent;
            if (grandparent == null) {
                break;
            }
            if (node.parent == grandparent.left) {
                Node uncle = grandparent.right;
                if (uncle.color == RED) {
                    node.parent.color = BLACK;
                    uncle.color = BLACK;
                    grandparent.color = RED;
                    node = grandparent;
                } else {
                    if (node == node.parent.right) {
                        node = node.parent;
                        leftRotate(node);
                    }
                    node.parent.color = BLACK;
                    grandparent.color = RED;
                    rightRotate(grandparent);
                }
            } else {
                Node uncle = grandparent.left;
                if (uncle.color == RED) {
                    node.parent.color = BLACK;
                    uncle.color = BLACK;
                    grandparent.color = RED;
                    node = grandparent;
                } else {
                    if (node == node.parent.left) {
                        node = node.parent;
                        rightRotate(node);
                    }
                    node.parent.color = BLACK;
                    grandparent.color = RED;
                    leftRotate(grandparent);
                }
            }
        }
        root.color = BLACK;
    }

    private void leftRotate(Node node) {
        Node right = node.right;
        node.right = right.left;
        if (right.left != nil) {
            right.left.parent = node;
        }
        right.parent = node.parent;
        if (node.parent == nil) {
            root = right;
        } else if (node == node.parent.left) {
            node.parent.left = right;
        } else {
            node.parent.right = right;
        }
        right.left = node;
        node.parent = right;
    }

    private void rightRotate(Node node) {
        Node left = node.left;
        node.left = left.right;
        if (left.right != nil) {
            left.right.parent = node;
        }
        left.parent = node.parent;
        if (node.parent == nil) {
            root = left;
        } else if (node == node.parent.right) {
            node.parent.right = left;
        } else {
            node.parent.left = left;
        }
        left.right = node;
        node.parent = left;
    }

    private void inorder(Node node, List<Integer> result) {
        if (node == nil) {
            return;
        }
        inorder(node.left, result);
        result.add(node.value);
        inorder(node.right, result);
    }

    private static final class Node {
        private final int value;
        private boolean color;
        private Node left;
        private Node right;
        private Node parent;

        private Node(int value, boolean color) {
            this.value = value;
            this.color = color;
        }
    }

    public static void main(String[] args) {
        int[] sample = {42, 7, 19, 3, 99, 1};
        int target = 19;
        RedBlackTree tree = new RedBlackTree();
        for (int value : sample) {
            tree.insert(value);
        }

        System.out.println("Red-Black Tree");
        System.out.println("input  : [42, 7, 19, 3, 99, 1]");
        System.out.println("inorder: " + tree.inorder());
        System.out.println("target : " + target);
        System.out.println("found  : " + tree.contains(target));
    }
}
